files/borgback-adfc-final.sh

```diff
@@ -0,0 +1,116 @@
+############ AB HIER KEINE ANPASSUNGEN MEHR ERFORDERLICH! #################################
+###########################################################################################
+
+
+# ---------- Einlesen der verbundenen Backup-Clients / Ermitteln der IP -------------------
+
+# Ist kein Client verbunden wird auch nichts unternommen
+if [ "$(ls -A $CL_ONLINE_DIR)" ]; then
+   MACHE="WEITER"
+else
+   exit 1
+fi
+
+# Einlesen aller Verbindungen
+ALLCLIENTS=()
+for FILES in $CL_ONLINE_DIR/*; do
+    ALLCLIENTS[${#ALLCLIENTS[@]}]="$FILES"
+done
+
+# Alle in CL_ONLINE_DIR vorhandenen Verbindungen testen, wenn Offline löschen. Nur 192er-Netz zulässig
+for FILES in "${ALLCLIENTS[@]}"; do
+    TEST_IP=$(echo "$FILES" | cut -d '=' -f 2)
+    if [ $(echo "$TEST_IP" | cut -d '.' -f 1) -ne 192 ]; then
+       rm $FILES
+    fi
+    ping -c 3 $TEST_IP 2>&1> /dev/null
+    RESULT=$?
+    if [ $RESULT == 0 ]; then
+       IPADDR=$TEST_IP
+       ONLINE_FILE=$FILES
+    else
+       rm $FILES
+    fi
+done
+
+# Falls keine Verbindung funktionierte ist das Verzeinis leer. Dann Abbruch!
+if [ "$(ls -A $CL_ONLINE_DIR)" ]; then
+   MACHE="WEITER"
+else
+   exit 2
+fi
+
+
+
+# ---------- Backup-Routine ----------------------------------------------------------------
+
+# Repo-Init legt beim erstmaligen Aufruf ein Repository an. Weitere Aufrufe beenden sich ohne Funktion oder Fehler (Rückgabewert=2 spielt hier keine Rolle).
+BORG_PASSPHRASE="$SECRET" borg init --encryption=repokey --show-rc $BORG_USER@$IPADDR:$REPO_PATH/$REPO_NAME 2>&1
+BORG_PASSPHRASE="$SECRET" borg create -v -s -p -C lz4 --show-version --list --show-rc $BORG_USER@$IPADDR:$REPO_PATH/$REPO_NAME::$ARCHIVE_NAME $SOURCE_DIR --exclude $SOURCE_EXCL $LOGFILE 2>&1
+
+
+
+# ---------- Backup-Protokolle per E-Mail versenden ----------------------------------------
+
+# Auf dem Backup-Client wird der letze Sicherungsstatus im borguser-Homedir hinterlegt. Der alte Wert wird zuvor gelöscht. Für Problemanalyse evtl. hilfreich
+ssh $BORG_USER@$IPADDR -t "rm ~/BACKUP_*" 2>&1
+
+# E-Mail bei Erfolg oder Fehlschlag versenden
+if [ $? == 0 ]; then
+   ssh $BORG_USER@$IPADDR -t "touch ~/BACKUP_ERFOLGREICH" # 2>&1
+   sendemail -f $M_SND -s $M_SVR -xu $M_SND -xp $M_PWD -t $M_RCV -u "$M_SUB - Datensicherung O.K." -m $M_BDY -a $LOGFILE $M_ATT 2>&1
+   echo "Dateiname = $ONLINE_FILE"
+   rm $ONLINE_FILE
+else
+   ssh $BORG_USER@$IPADDR -t "touch ~/BACKUP_FEHLGESCHLAGEN" 2>&1
+   sendemail -f $M_SND -s $M_SVR -xu $M_SND -xp $M_PWD -t $M_RCV -u "$M_SUB - FEHLER BEI DATENSICHERUNG!!" -m $M_BDY -a $LOGFILE $M_ATT 2>&1
+   mv $ONLINE_FILE ${LOG_DIR}_FAILED
+   exit 9
+fi
+
+
+
+# ---------- Alte Backups & Logdaten aufräumen --------------------------------------------------------
+
+BORG_PASSPHRASE="$SECRET" borg prune $BORG_USER@$IPADDR:$REPO_PATH/$REPO_NAME --keep-within=1d --keep-daily=$PRUNE_D --keep-weekly=$PRUNE_W --keep-monthly=$PRUNE_M --keep-yearly=$PRUNE_Y
+
+find $LOG_DIR/* -mtime 30 -exec rm {} \;
```

Comment:
Dafür gibt es Logrotate. Und die Logdateien schon nach 30 Minuten zu löschen, finde ich zu früh.

Reply:
30 Minuten wäre in der Tat zu früh. Nach meiner Kenntnis erfolgt nach obiger Syntax aber erst nach 30 Tagen die Löschung (siehe Manpage):

   -mtime n
                 File's data was last modified n*24 hours ago.  See the  comments
                 for -atime to understand how rounding affects the interpretation
                 of file modification times.

Ich habe mir die Vorausetzungen und Beispiele zu 'logrotate' einmal angeschaut und empfinde sie als deutlich komplexer als meinen Einzeiler. Vielleicht verstehe ich aber auch etwas falsch.

Reply:
Du hast recht sind 30 Tage. Trotzdem würde ich Logrotate bevorzugen. Weil es eben genau dafür da ist und z.B. Logfiles auch komprimiert etc.

Etwas erfolgreiches nicht zu nutzen und lieber selbst zu programmieren ist in der Informatik selten Sinnvoll.

Außerdem wird dein Befehl jede Stunde ausgeführt. Logrotate nur einmal am Tag.

Es reicht in der Regel dafür eine Datei in /etc/logrotate.d zu erstellen, z.B. mit Inhalt:

```
/var/log/adfc-backup/* {
        daily
        missingok
        rotate 30
        compress
        delaycompress
        notifempty
}
```
(ungetestet)

Ein Verzeichnis was in $LOG_DIR angelegt wird, führt z.B. zu einer Fehlermeldung